### PR TITLE
Added Support for pycocotools on Windows

### PR DESCRIPTION
**Pull request by sarthakforwet**
Updated Linting of the code and added support for pycocotools build on Windows. 

PR from the discussion at [#1712(comment)](https://github.com/facebookresearch/detectron2/issues/1712#issuecomment-670996808).

### CODE REVIEW

File: PythonAPI/setup.py [hunk 1]
```diff
@@ -1,37 +1,47 @@
+"""To compile and install locally run "python setup.py build_ext --inplace".
+To install library to Python site-packages run "python setup.py build_ext install"
+"""
+import platform
+import numpy as np
```
Comment (ppwwyyxx) [inline PythonAPI/setup.py:5]: IIRC this line needs to go after `fetch_build_eggs` so that `numpy` can be installed automatically if not yet available

**Comment by sarthakforwet** [comment]
> The second step is to make sure a distribution can be made that can later be installed on windows.

I ran `python setup.py sdist` and the following is the output. It seems to be working to me..
```bash
$ python setup.py sdist
running sdist
running egg_info
writing pycocotools.egg-info\PKG-INFO
writing dependency_links to pycocotools.egg-info\dependency_links.txt
writing requirements to pycocotools.egg-info\requires.txt
writing top-level names to pycocotools.egg-info\top_level.txt
reading manifest file 'pycocotools.egg-info\SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'pycocotools.egg-info\SOURCES.txt'
warning: sdist: standard file not found: should have one of README, README.rst, README.txt, README.md

running check
warning: check: missing required meta-data: url

warning: check: missing meta-data: either (author and author_email) or (maintainer and maintainer_email) must be supplied

creating pycocotools-2.0.1
creating pycocotools-2.0.1\common
creating pycocotools-2.0.1\pycocotools
creating pycocotools-2.0.1\pycocotools.egg-info
copying files to pycocotools-2.0.1...
copying MANIFEST.in -> pycocotools-2.0.1
copying setup.py -> pycocotools-2.0.1
copying common\gason.cpp -> pycocotools-2.0.1\common
copying common\gason.h -> pycocotools-2.0.1\common
copying common\maskApi.c -> pycocotools-2.0.1\common
copying common\maskApi.h -> pycocotools-2.0.1\common
copying pycocotools\__init__.py -> pycocotools-2.0.1\pycocotools
copying pycocotools\_mask.c -> pycocotools-2.0.1\pycocotools
copying pycocotools\_mask.pyx -> pycocotools-2.0.1\pycocotools
copying pycocotools\coco.py -> pycocotools-2.0.1\pycocotools
copying pycocotools\cocoeval.py -> pycocotools-2.0.1\pycocotools
copying pycocotools\mask.py -> pycocotools-2.0.1\pycocotools
copying pycocotools.egg-info\PKG-INFO -> pycocotools-2.0.1\pycocotools.egg-info
copying pycocotools.egg-info\SOURCES.txt -> pycocotools-2.0.1\pycocotools.egg-info
copying pycocotools.egg-info\dependency_links.txt -> pycocotools-2.0.1\pycocotools.egg-info
copying pycocotools.egg-info\requires.txt -> pycocotools-2.0.1\pycocotools.egg-info
copying pycocotools.egg-info\top_level.txt -> pycocotools-2.0.1\pycocotools.egg-info
Writing pycocotools-2.0.1\setup.cfg
Creating tar archive
removing 'pycocotools-2.0.1' (and everything under it)
```

**Comment by sarthakforwet** [comment]
I've also run `python setup.py build install` on both Windows and Linux and its working fine.

**Comment by ppwwyyxx** [comment]
Cool! could you try `pip install output.tar` on the output tar file from `python setup.py sdist` ?

File: PythonAPI/setup.py (hunk 1 above)
Comment (sarthakforwet) [reply PythonAPI/setup.py:5]: We can include installation of `numpy` as a dependency to install `pycocotools` as `cython` is. WDYT @ppwwyyxx? This improves the linting and code structure.

Or should I change the position of the line back after `fetch_build_eggs`?

**Comment by ppwwyyxx** [comment]
Just added CI for the linux build: https://github.com/ppwwyyxx/cocoapi/pull/3
Would be nice to have it for windows as well to verify the changes - or it can also be verified later on the CI of detectron2.

File: PythonAPI/setup.py (hunk 1 above)
Comment (ppwwyyxx) [reply PythonAPI/setup.py:5]: `numpy` is a dependency to execute this `setup.py` file (used in extension declaration). Therefore declaring it as a dependency in `setup()` would be too late and IIRC it won't work. Putting it after `fetch_build_eggs` should work

**Comment by sarthakforwet** [comment]
> Cool! could you try pip install output.tar on the output tar file from python setup.py sdist 

I've tried but it is giving error that `maskAPI.c` file is not present.

I've tried `python setup.py sdist` /  `python setup.py sdist --formats=zip`. Both did not work.

PS: I've also tried building pycocotools after changing the numpy import line and its also failing at pip install .

**Comment by ppwwyyxx** [comment]
You can untar the file to see what it looks like inside. My guess (according to how it works on Linux) is that inside the tar file the common directory is at "./common", not "../common". But in `setup.py` it's using "../common" for windows

**Comment by sarthakforwet** [comment]
@ppwwyyxx. It finally worked with `pip install pycocotools.zip`. You were right, it was `./common` and not `../common`.

File: PythonAPI/setup.py (hunk 1 above)
Comment (sarthakforwet) [reply PythonAPI/setup.py]: When building pycocotools in a new environment, it requires `cython` to be installed. Should the file be changed to build `cython` if not already installed?

File: PythonAPI/setup.py [hunk 2]
```diff
@@ -12,26 +13,36 @@
...
+            extra_compile_args=[],
+        )
+    ]
+else:
+    ext_modules = [
+        Extension(
+            'pycocotools._mask',
+            sources=['./common/maskApi.c', 'pycocotools/_mask.pyx'],
+            include_dirs=[np.get_include(), './common'],
+            extra_compile_args=['-Wno-cpp', '-Wno-unused-function', '-std=c99'],
```
Comment (ppwwyyxx) [inline PythonAPI/setup.py:31]: if this line is the only difference, you can move the `if windows ...` to the right hand side of this line

File: PythonAPI/setup.py (hunk 1 above)
Comment (ppwwyyxx) [reply PythonAPI/setup.py]: When it's missing, `dist.Distribution().fetch_build_eggs` takes care of installing cython before the real installation starts. This process works on Linux but don't know if things are the same on windows. If this cannot work maybe we'll have to tell users to manually install cython first

Comment (sarthakforwet) [reply PythonAPI/setup.py]: > This process works on Linux but don't know if things are the same on windows. 

You are right, this works on Linux. But it does not work on Windows. We need to add description to install `cython` prior to the installation of pycocotools.

File: PythonAPI/setup.py (hunk 2 above)
Comment (sarthakforwet) [reply PythonAPI/setup.py:31]: >  I'm not sure it works the same on windows. Please double-check the whole process from the beginning when you finalize the PR

I executed the complete process again and its completely working fine over both Windows and Linux with a prior installation of Cython on Windows.  

> if this line is the only difference, you can move the if windows ... to the right hand side of this line

Done! Thanks!

**Comment by ppwwyyxx** [comment]
Thanks! I'll create an alpha release on pypi for testing

**Comment by sarthakforwet** [comment]
> Thanks! I'll create an alpha release on pypi for testing

Thanks @ppwwyyxx! I think now we could build detectron2 for `Windows` more easily.

**Comment by ppwwyyxx** [comment]
https://pypi.org/project/pycocotools/2.0.2a1/#files
use this for detectron2 build - I can turn it to stable release later when everything is working

**Pull request merged**